### PR TITLE
replace all app:action string permissions with tuple constants

### LINKED ISSUE #3572: Replace permission strings with constants
We shouldn't be passing around random strings to define what permissions are needed; everything should be a constant.

### PR DESCRIPTION
**Pull request by eviljeff**
fixes #3572

### CODE REVIEW

File: src/olympia/access/tests.py [hunk 1]
```diff
@@ -60,7 +60,7 @@ def test_match_rules():
 
 def test_anonymous_user():
     fake_request = req_factory_factory('/')
-    assert not action_allowed(fake_request, amo.FIREFOX, 'Admin:%')
```
Comment (eviljeff) [inline src/olympia/access/tests.py]: this never worked, fwiw

File: src/olympia/access/acl.py [hunk 2]
```diff
@@ -40,7 +38,8 @@ def action_allowed_user(user, app, action):
         return False
 
     return any(
-        match_rules(group.rules, app, action) for group in user.groups_list)
+        match_rules(group.rules, permission.app, permission.action)
+        for group in user.groups_list)
```
Comment (EnTeQuAk) [inline src/olympia/access/acl.py]: I actually wonder if we should enforce that a permission **has to be** in `constants.permissions` to make sure we don't allow any none-constants?

File: src/olympia/constants/permissions.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+from collections import namedtuple
+
+
+AclPermission = namedtuple('AclPermission', 'app, action')
+
```
Comment (EnTeQuAk) [inline src/olympia/constants/permissions.py]: Improvement idea: Allow a `help_text` to be set (all the comments for every constant) so that we are able to show them in the admin at some point?

File: src/olympia/constants/permissions.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
+from collections import namedtuple
+
+
+AclPermission = namedtuple('AclPermission', 'app, action')
+
+# Special null rule.
+NONE = AclPermission('None', 'None')
+
```
Comment (EnTeQuAk) [inline src/olympia/constants/permissions.py]: This appears only to be used for testing is that correct?

File: src/olympia/constants/permissions.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+AclPermission = namedtuple('AclPermission', 'app, action')
+
+# Special null rule.
+NONE = AclPermission('None', 'None')
+
+# Admin super powers.  Very few users will have this permission (2-3)
+ADMIN = AclPermission('Admin', '%')
+
+# Can view admin tools.
+ADMINTOOLS = AclPermission('AdminTools', 'View')
```
Comment (EnTeQuAk) [inline src/olympia/constants/permissions.py]: Let's be consistent about this maybe and call the constant `ADMIN_TOOLS`? similar with the others below?

File: src/olympia/constants/permissions.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+# Can edit all collections.
+COLLECTIONS_EDIT = AclPermission('Collections', 'Edit')
+
+# Can view statistics for all addons, regardless of privacy settings.
+STATS = AclPermission('Stats', 'View')
+# Can view collection statistics.
+COLLECTIONSTATS = AclPermission('CollectionStats', 'View')
+# Can view revenue statistics.
+REVENUESTATS = AclPermission('RevenueStats', 'View')
```
Comment (EnTeQuAk) [inline src/olympia/constants/permissions.py]: I'd explicitly name this `REVENUE_STATS_VIEW`, if we have a `_EDIT` postfix we should have it's `_VIEW` (or `_READ` for that matter) pendent.

File: src/olympia/access/acl.py (hunk 2 above)
Comment (eviljeff) [inline src/olympia/access/acl.py]: Enforce how?  Run-time `assert`?  Or just a test? (We already have a test that tests for a random permission)

File: src/olympia/constants/permissions.py (hunk 3 above)
Comment (eviljeff) [inline src/olympia/constants/permissions.py]: I'll do it if you promise to write the patch to add them to admin 😛

File: src/olympia/constants/permissions.py (hunk 4 above)
Comment (eviljeff) [inline src/olympia/constants/permissions.py]: Yes... it appears so.  I'll remove it from constants.

File: src/olympia/constants/permissions.py (hunk 5 above)
Comment (eviljeff) [inline src/olympia/constants/permissions.py]: I was aiming for a consistent `THEAPPNAME_THEACTIONNAME` for `TheAppName:TheActionName`, though I'm aware that's not consistent with python constant naming.  The extra `_`s means its not always possible to know if `FOO_BAR_BOO` is `FooBar:Boo` or `Foo:BarBoo` from just the constant name.  Is `ADMIN_TOOLS__VIEW` too wordy?

File: src/olympia/constants/permissions.py (hunk 6 above)
Comment (eviljeff) [inline src/olympia/constants/permissions.py]: I was aiming for less words where possible, so assuming `View`, but I can change it to be explicit.

File: src/olympia/access/acl.py (hunk 2 above)
Comment (EnTeQuAk) [inline src/olympia/access/acl.py]: run time `assert` yeah… But I think a unittest is fine too.

File: src/olympia/constants/permissions.py (hunk 3 above)
Comment (EnTeQuAk) [inline src/olympia/constants/permissions.py]: doh!

File: src/olympia/constants/permissions.py (hunk 4 above)
Comment (EnTeQuAk) [inline src/olympia/constants/permissions.py]: With https://github.com/mozilla/addons-server/pull/5178#discussion_r112623186 you shouldn't be able to remove it… imho… maybe just document what it's being used for is fine.

File: src/olympia/constants/permissions.py (hunk 6 above)
Comment (EnTeQuAk) [inline src/olympia/constants/permissions.py]: maybe cc @diox for an opinion but I prefer explicit names over an assumed meaning. It's a nitpick but for people not aware of these implicit meanings I think it's easier to read our code and reason about permissions.

File: src/olympia/constants/permissions.py (hunk 5 above)
Comment (EnTeQuAk) [inline src/olympia/constants/permissions.py]: I don't think the constant names should be consistent with how they end up in the admin or database. So imho, `ADMIN_TOOLS_VIEW` is perfect for me, note only one underline. The representation in the database should be irrelevant for their usage in code so let's make them consistent with python naming.

File: src/olympia/access/acl.py (hunk 2 above)
Comment (eviljeff) [inline src/olympia/access/acl.py]: https://github.com/mozilla/addons-server/pull/5178/commits/e55e0f82ed4dc2c4d4fd1e619558628945bc4cb9

File: src/olympia/constants/permissions.py (hunk 4 above)
Comment (eviljeff) [inline src/olympia/constants/permissions.py]: https://github.com/mozilla/addons-server/pull/5178/commits/e55e0f82ed4dc2c4d4fd1e619558628945bc4cb9

File: src/olympia/constants/permissions.py (hunk 5 above)
Comment (eviljeff) [inline src/olympia/constants/permissions.py]: https://github.com/mozilla/addons-server/pull/5178/commits/2060f032341c19452e3036ae828be2ef51cb20bc

File: src/olympia/constants/permissions.py (hunk 6 above)
Comment (eviljeff) [inline src/olympia/constants/permissions.py]: https://github.com/mozilla/addons-server/pull/5178/commits/c3abecfa5902f3bb66ab247d04f1bc6cf37a79ab
